### DISCUSSION TITLE
input type number with leading zero

### DISCUSSION THREAD

**Discussion by DanielGennat** [neuefische/questions#202] [Web] [answered]
Hi there,
how can I create an input field of type number with leading zero if the current value is < 10?

<img width="905" alt="Bildschirmfoto 2022-10-04 um 11 50 26" src="https://user-images.githubusercontent.com/109812125/193789781-7e96e940-38ac-4f3d-a393-dddfd75ad029.png">

Here is my [PR](https://github.com/DanielGennat/capstone-project/pull/12).

Thanks in advance :)

**Comment by NikRoe** [comment, marked as answer]
Hey Daniel,
you would need to work with controlled inputs, a state and an onChange event in order for this to happen.
It could look similar to this (however, in this example, all inputs share the same state and this won't be what you want):
```js
export default function EditTimerChainForm({ timerChain, setTimerChain }) {
  const [time, setTime] = useState(0);
//...
            <InputAndLabelWrapper>
              <Input
                type="number"
                id={`seconds of ${timer.id}`}
                name={`seconds of ${timer.id}`}
                min="0"
                max="59"
                value={time < 10 ? `0${time}` : time}
                onChange={(event) => setTime(event.target.value)}
                required
              />
              <label htmlFor={`seconds of ${timer.id}`}> sec.</label>
            </InputAndLabelWrapper>
//...
```

To make it work properly in your App, you would need to update the state whenever there is a change in an input field :)

**Reply by DanielGennat** [reply]
Hi Niklas, bei mir ergibt sich jetzt das nächste Problem: Die leading Zero bleibt stehen, wenn ich zb bei 00  020 eingebe. Ich habe es an zwei verschiedenen Stellen mit Slicing probiert, aber es funktioniert nicht so wie es soll.

[PR](https://github.com/DanielGennat/capstone-project/pull/12)

**Reply by DanielGennat** [reply]
Der value wird korrekt gesliced in der Console geloggt, aber nicht mit dem gleichen Wert im State übernommen (und vermutlich deshalb auch nicht korrekt im input angezeigt) ... Warum? :)
Hier ein Beispiel, wo ich ins zweite Sekundenfeld bei 00 -> 010 eintrage:
![Bildschirmfoto 2022-10-06 um 13 07 04](https://user-images.githubusercontent.com/109812125/194297848-49784e03-33bb-4aee-aa5f-22d92c1e676d.png)
und hier die `changeTimer` Funktion:
<img width="876" alt="Bildschirmfoto 2022-10-06 um 13 08 46" src="https://user-images.githubusercontent.com/109812125/194298116-85e0b684-d3dc-4ffd-a951-d1b3e461f451.png">

[PullRequest](https://github.com/DanielGennat/capstone-project/pull/12)

**Reply by NikRoe** [reply]
Hi Daniel,
eine 0 lässt sich nicht vor Zahlen schreiben, da dies vom [Zahlensystem](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Deprecated_octal) nicht unterstützt wird.

Deswegen übernimmt dein State die vorgestellte 0 nicht.

Du kannst dich dem Hack mit dem `type="decimal"` bedienen. Da dies aber nicht wirklich im Sinne des Erfinders ist, würde ich nicht dazu tendieren.

Wie wäre es denn mit einem Dropdown, aus dem du in 5er/10er-Schritten Werte auswählen kannst?

**Reply by DanielGennat** [reply]
ja, `type="decimal"` funktioniert bei mir so lange bis man etwas anderes als eine Zahl eingibt

**Reply by DanielGennat** [reply]
Es wäre eventuell sehr unbefriedigend, aber ich probiere es mal

**Reply by DanielGennat** [reply]
Wie kann ich denn ein Drop-Down von 00 bis 59 erstellen ohne 60 Zeilen Code dafür zu schreiben?

**Reply by DanielGennat** [reply]
das Drop-Down IST sehr befriedigend, da es jetzt auch mit leading zero funktioniert. :)
Vielen Dank für deine Unterstützung!
Hier die Lösung:

Array mit loop für die Optionen von 0 bis 59:
```
const timeOptions = [];
  for (let i = 0; i <= 59; i++) {
    timeOptions.push(i);
  }
```
select mit  onChange und map für die Optionen:
```
<Select
                id="seconds"
                value={newTimer.seconds}
                onChange={(event) =>
                  changeTimerChain(
                    event,
                    newTimer.id,
                    newTimerChain,
                    setNewTimerChain
                  )
                }
              >
                {timeOptions.map((option) => {
                  return (
                    <option key={option} value={option}>
                      {option < 10 ? `0${option}` : option}
                    </option>
                  );
                })}
              </Select>
              <Label htmlFor="seconds"> sec.</Label>
```
und die changeTimerChain Funktion:
```
export function changeTimerChain(event, id, newTimerChain, setNewTimerChain) {
    setNewTimerChain(
      newTimerChain.map((newTimer) => {
        if (newTimer.id === id) {
          return { ...newTimer, seconds: Number(event.target.value) };
        }
        return newTimer;
      })
    );
```
